crates/uv-keyring/src/windows.rs

```diff
@@ -0,0 +1,768 @@
+                comment: format!("{user}@{service}:{target} (keyring v{VERSION})"),
+            }
+        } else {
+            Self {
+                // Note: default concatenation of user and service name is
+                // used because windows uses target_name as sole identifier.
+                // See the module docs for more rationale.  Also see this issue
+                // for Python: https://github.com/jaraco/keyring/issues/47
+                //
+                // Note that it's OK to have an empty user or service name,
+                // because the format for the target name will not be empty.
+                // But it's certainly not recommended.
+                username: user.to_string(),
+                target_name: format!("{user}.{service}"),
+                target_alias: String::new(),
+                comment: format!("{user}@{service}:{user}.{service} (keyring v{VERSION})"),
+            }
+        };
+        credential.validate_attributes(None, None)?;
+        Ok(credential)
+    }
+}
+
+/// The builder for Windows Generic credentials.
+pub struct WinCredentialBuilder;
+
+/// Returns an instance of the Windows credential builder.
+///
+/// On Windows, with the default feature set,
+/// this is called once when an entry is first created.
+pub fn default_credential_builder() -> Box<CredentialBuilder> {
+    Box::new(WinCredentialBuilder {})
+}
+
+impl CredentialBuilderApi for WinCredentialBuilder {
+    /// Build a [`WinCredential`] for the given target, service, and user.
+    fn build(&self, target: Option<&str>, service: &str, user: &str) -> Result<Box<Credential>> {
+        Ok(Box::new(WinCredential::new_with_target(
+            target, service, user,
+        )?))
+    }
+
+    /// Return the underlying builder object with an `Any` type so that it can
+    /// be downgraded to a [`WinCredentialBuilder`] for platform-specific processing.
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+}
+
+fn extract_password(credential: &CREDENTIALW) -> Result<String> {
+    let mut blob = extract_secret(credential)?;
+    // 3rd parties may write credential data with an odd number of bytes,
+    // so we make sure that we don't try to decode those as utf16
+    if blob.len() % 2 != 0 {
+        return Err(ErrorCode::BadEncoding(blob));
+    }
+    // This should be a UTF-16 string, so convert it to
+    // a UTF-16 vector and then try to decode it.
+    let mut blob_u16 = vec![0; blob.len() / 2];
+    LittleEndian::read_u16_into(&blob, &mut blob_u16);
+    let result = match String::from_utf16(&blob_u16) {
+        Err(_) => Err(ErrorCode::BadEncoding(blob)),
+        Ok(s) => {
+            // we aren't returning the blob, so clear it
+            blob.zeroize();
+            Ok(s)
+        }
+    };
+    // we aren't returning the utf16 blob, so clear it
+    blob_u16.zeroize();
+    result
+}
+
+#[allow(clippy::unnecessary_wraps)]
+fn extract_secret(credential: &CREDENTIALW) -> Result<Vec<u8>> {
```

Comment:
Agreed that it's not worth propagating `unsafe` everywhere, since these are module-private functions in vendored code.